Fix: Cannot use 'type' => 'file' on HelperOptions on PS 9.0.x

Codencode opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | see #38133
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | see #38133
| UI Tests          | https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/15438873126
| Fixed issue or discussion?     | Resolves #38133 Resolves #38321
| Related PRs       | #38006
| Sponsor company   | @Codencode

Codencode committed:
We need to check $this->getContext()->smarty->getTemplateDir(0) before $this->getContext()->smarty->getTemplateDir(1).

kpodemski commented:
I didn't check it but it seems that you change the priority of the overrides with this change. I might be wrong, but that's my first thought.

Codencode commented:
> I didn't check it but it seems that you change the priority of the overrides with this change. I might be wrong, but that's my first thought.

Then I need to test with an override.
Thank you!

Codencode commented:
@kpodemski,

I conducted some checks, and in PrestaShop version 9.x.x, from what I've observed, the instruction `$this->getContext()->smarty->getTemplateDir(1)` always returns `null`. This is because here: https://github.com/PrestaShop/PrestaShop/blob/8ec9a5d9717fed5c36b71b8394cfe31bfe645d4f/src/PrestaShopBundle/Controller/Admin/LegacyController.php#L119 only one path is set.
In contrast, in version 8.2.0, the override path is also passed, and it's done here: https://github.com/PrestaShop/PrestaShop/blob/965713ab343def9d3058df168219790cbe968df5/classes/controller/AdminController.php#L484-L487

From what I can see, it seems as though the override mechanism has been "bypassed" in version 9.

If that's the case, in theory, my modification could work, considering it's a temporary fix for a functionality (the use of Helpers for developing admin areas) that I believe will gradually be phased out in favor of Symfony.

I hope I've been clear enough.

jolelievre commented:
Interesrting comment @Codencode 
It was not planned to remove this override feature in V9, so it looks like a side effect, maybe you can try and add this extra template dir in the new `LegacyController` SF controller, if it doesn't trigger some unexpected error with the UI tests then it means it's safe to keep both potential folders, and then your initial fix can be removed so we ensure the same priority for overrides

Codencode committed:
Restored the possibility to override controllers from the old Prestashop controller system

By restoring the override capability, the modification to the method classes/helper/HelperUploaderCore::getTemplateFile() that I introduced in the previous commit is no longer needed.

Codencode commented:
@jolelievre,

I reintroduced the ability to perform overrides by modifying the LegacyController.
This change alone resolves the issue, so I removed the modification I had introduced in the previous commit.

idnovate commented:
@kpodemski Any possibility to merge this PR before PS 9 final release? It's breaking some modules. Thanks!

kpodemski commented:
@idnovate 

I'm trying my best. I'll get back to you.

SiraDIOP reviewed (commented):
Hello @Codencode,

Thanks for the PR,

I tested the pr for the issue https://github.com/PrestaShop/PrestaShop/issues/38133

https://github.com/user-attachments/assets/7321cd1c-33e1-4de9-ac21-e3f75e7e5048

It's OK for me^^
For the second issue https://github.com/PrestaShop/PrestaShop/issues/38321
I have an error when installing the module: [dummymodule.zip](https://github.com/user-attachments/files/20592704/dummymodule.zip)
<img width="1423" alt="Capture d’écran 2025-06-04 à 15 52 00" src="https://github.com/user-attachments/assets/f9c559f8-1eb5-4bb2-acd3-d73d1c539216" />
Did i miss something?
Thanks in advance for your feedback ^^

kpodemski commented:
@SiraDIOP could you try to clear the cache of your PrestaShop, go to the Modules, and try to install module again?

SiraDIOP reviewed (commented):
Hello @kpodemski, @Codencode 

I cleared the cache and it works. Thanks 

It's Ok for me :)

Uploading Enregistrement de l’écran 2025-06-05 à 10.32.53.mov…


I'm just waiting for the automated tests to be green and I'll validate the PR
